lib/shared-types.d.ts

```diff
@@ -0,0 +1,297 @@
+   * and `returnState` is switched to.
+   * If the result is `nok`, the attempt failed,
+   * so we revert to the original state, and `bogusState` is used.
+   */
+  attempt: (
+    constructInfo:
+      | Construct
+      | Construct[]
+      | Record<CodeAsKey, Construct | Construct[]>,
+    returnState: State,
+    bogusState?: State
+  ) => (code: Code) => void
+
+  /**
+   * Interrupt is used for stuff right after a line of content.
+   */
+  interrupt: (
+    constructInfo:
+      | Construct
+      | Construct[]
+      | Record<CodeAsKey, Construct | Construct[]>,
+    ok: Okay,
+    nok?: NotOkay
+  ) => (code: Code) => void
+
+  check: (
+    constructInfo:
+      | Construct
+      | Construct[]
+      | Record<CodeAsKey, Construct | Construct[]>,
+    ok: Okay,
+    nok?: NotOkay
+  ) => (code: Code) => void
+
+  /**
+   * Lazy is used for lines that were not properly preceded by the container.
+   */
+  lazy: (
+    constructInfo:
+      | Construct
+      | Construct[]
+      | Record<CodeAsKey, Construct | Construct[]>,
+    ok: Okay,
+    nok?: NotOkay
+  ) => void
+}
+
+/**
+ * A state function should return another function: the next state-as-a-function to go to.
+ *
+ * But there is one case where they return void: for the eof character code (at the end of a value)
+ * The reason being: well, there isn’t any state that makes sense, so void works well. Practically
+ * that has also helped: if for some reason it was a mistake, then an exception is throw because
+ * there is no next function, meaning it surfaces early.
+ */
+export type State = (code: number) => State | void
+
+/**
+ *
+ */
+export type Okay = State
+
+/**
+ *
+ */
+export type NotOkay = State
+
+/**
+ *
+ */
+export interface Tokenizer {
+  previous: Code
+  events: Event[]
+  parser: Parser
+  sliceStream: (token: Token) => Chunk[]
+  sliceSerialize: (token: Token) => string
+  now: () => Point
+  defineSkip: (value: Point) => void
+  write: (slice: Chunk[]) => Event[]
+}
+
+export type Resolve = (events: Event[], context: Tokenizer) => Event[]
+
+export type Tokenize = (context: Tokenizer, effects: Effects) => State
+
+export interface Construct {
+  tokenize: Tokenize
+  partial?: boolean
+  resolve?: Resolve
+  resolveTo?: Resolve
+  resolveAll?: Resolve
+  concrete?: boolean
+  interruptible?: boolean
+  lazy?: boolean
+}
+
+/**
+ *
+ */
+export interface Parser {
+  constructs: Record<CodeAsKey, Construct | Construct[]>
+  content: (from: Point) => Tokenizer
+  document: (from: Point) => Tokenizer
+  flow: (from: Point) => Tokenizer
+  string: (from: Point) => Tokenizer
+  text: (from: Point) => Tokenizer
+  defined: string[]
+}
+
+/**
+ *
+ */
+export interface TokenizerThis {
+  events: Event[]
+  interrupt: unknown
+  lazy: unknown
+  containerState: {
+    marker: number
+    type: Type
+    initialBlankLine: unknown
+    size: number
+    _closeFlow: unknown
+    furtherBlankLines: unknown
```

Comment:
do these unknowns have more concrete types?

Reply:
extensions can put whatever they want on there, although `_closeFlow` is a symbol to make lists work. `interrupt` and `lazy` are optional booleans. What if we allow `containerState` to be an object that allows any key/value?